README
Simple Python Library to control Citrix Netscaler 9.2+ load balancers with NITRO API.
What we're trying here is to implement Citrix NITRO API SDK (that's available for Java and C#) in Python.

You can find a simple command-line interface using this library in contrib/.
Feedback/comments welcome.

Very basic usage::

import time
from nsnitro import *

nitro = NSNitro('localhost', 'api_user', 'api_user')
nitro.login()

# add server test
addserver = NSServer()
addserver.set_name("mp-nitroserver")
addserver.set_ipaddress("10.32.110.99")
NSServer.add(nitro, addserver)

server = NSServer()
server.set_name("mp-nitroserver")
server = server.get(nitro, server)
print server.get_name() + ": " + server.get_state()

# disable server test
server = NSServer()
server.set_name("mp-nitroserver")
NSServer.disable(nitro, server)

time.sleep(2)

server = NSServer()
server.set_name("mp-nitroserver")
server = server.get(nitro, server)
print server.get_name() + ": " + server.get_state()


# enable server test
server = NSServer()
server.set_name("mp-nitroserver")
NSServer.enable(nitro, server)

time.sleep(2)

server = NSServer()
server.set_name("mp-nitroserver")
server = server.get(nitro, server)
print server.get_name() + ": " + server.get_state()

# add service test

addservice = NSService()
addservice.set_name("aurora_testnitroadd")
addservice.set_servername("mp-nitroserver")
addservice.set_servicetype("HTTP")
addservice.set_port(11111)
NSService.add(nitro, addservice)

# add lbvserver test
lbvserver = NSLBVServer()
lbvserver.set_name("nitro_lbvserver_test")
lbvserver.set_ipv46("10.32.110.55")
lbvserver.set_port(11111)
lbvserver.set_clttimeout(180)
lbvserver.set_persistencetype("NONE")
lbvserver.set_servicetype("HTTP")
NSLBVServer.add(nitro, lbvserver)
#
print "LB vserver added"

# bind service to lbvserver test
lbbinding = NSLBVServerServiceBinding()
lbbinding.set_name("nitro_lbvserver_test")
lbbinding.set_servicename("aurora_testnitroadd")
lbbinding.set_weight(40)
NSLBVServerServiceBinding.add(nitro, lbbinding)

print "Binding added"

lbbinding = NSLBVServerServiceBinding()
lbbinding.set_name("nitro_lbvserver_test")
lbbindings = NSLBVServerServiceBinding.get(nitro, lbbinding)

for lbb in lbbindings:
        print "sgn: " + lbb.get_servicegroupname()

# delete binding test
lbbinding = NSLBVServerServiceBinding()
lbbinding.set_name("nitro_lbvserver_test")
lbbinding.set_servicename("aurora_testnitroadd")
NSLBVServerServiceBinding.delete(nitro, lbbinding)

print "Binding removed"

# delete lbvserver test
lbvserver = NSLBVServer()
lbvserver.set_name("nitro_lbvserver_test")
NSLBVServer.delete(nitro, lbvserver)

# get service test

service = NSService()
service.set_name("aurora_testnitroadd")
service = service.get(nitro, service)
print service.get_name() + ": " + service.get_svrstate()
print service.get_name() + ": %s %s" % (service.get_port(), service.get_useproxyport())

#update service test
updateservice = NSService()
updateservice.set_name("aurora_testnitroadd")
updateservice.set_comment("test comment")
updateservice.set_useproxyport("NO")
NSService.update(nitro, updateservice)

# get service test

service = NSService()
service.set_name("aurora_testnitroadd")
service = service.get(nitro, service)
print service.get_name() + ": " + service.get_svrstate()
print service.get_name() + ": %s %s %s" % (service.get_port(), service.get_comment(), service.get_useproxyport())

# disable service test

disservice = NSService()
disservice.set_name("aurora_testnitroadd")
NSService.disable(nitro, disservice)

service = NSService()
service.set_name("aurora_testnitroadd")
service = service.get(nitro, service)
print service.get_name() + ": " + service.get_svrstate()

# enable service test

enservice = NSService()
enservice.set_name("aurora_testnitroadd")
NSService.enable(nitro, enservice)

service = NSService()
service.set_name("aurora_testnitroadd")
service = service.get(nitro, service)
print service.get_name() + ": " + service.get_svrstate()


# rename service test

renservice = NSService()
renservice.set_name("aurora_testnitroadd")
renservice.set_newname("aurora_testnitroadd_rename")
NSService.rename(nitro, renservice)

# rename service back test

renservice = NSService()
renservice.set_name("aurora_testnitroadd_rename")
renservice.set_newname("aurora_testnitroadd")
NSService.rename(nitro, renservice)

# delete service test

delservice = NSService()
delservice.set_name("aurora_testnitroadd")
NSService.delete(nitro, delservice)

try:
        service = NSService()
        service.set_name("aurora_testnitroadd")
        service = service.get(nitro, service)
        print service.get_name() + ": " + service.get_svrstate()
except NSNitroError, e:
        print e.message


# delete server test
delserver = NSServer()
delserver.set_name("mp-nitroserver")
NSServer.delete(nitro, delserver)


try:
        server = NSServer()
        server.set_name("mp-nitroserver")
        server = server.get(nitro, server)
        print server.get_name() + ": " + server.get_state()
except NSNitroError, e:
        print e.message

#add system cmd policy
cmdpol = NSSystemCMDPolicy()
cmdpol.set_action('ALLOW')
cmdpol.set_policyname('nitro-cmd-policy')
cmdpol.set_cmdspec('show hardware')
NSSystemCMDPolicy.add(nitro,cmdpol)

#update system cmd policy
cmdpol = NSSystemCMDPolicy()
cmdpol.set_action('DENY')
cmdpol.set_policyname('nitro-cmd-policy')
cmdpol.set_cmdspec('show hardware')
NSSystemCMDPolicy.update(nitro,cmdpol)

#delete system cmd policy
cmdpol = NSSystemCMDPolicy()
cmdpol.set_policyname('nitro-cmd-policy')
NSSystemCMDPolicy.delete(nitro,cmdpol)


debian/changelog
nsnitro (1.0.35) unstable; urgency=low

  * Fixed a typo in init

 -- vlazarenko <vlazarenko@squeeze.par>  Tue, 1 Dec 2015 11:41:05 +0100

nsnitro (1.0.34) unstable; urgency=low

  * Added new resources:
  *   Auth LDAP Action (LDAP server(s) definition)
  *   Auth LDAP Policy
  *   System Global LDAP Ploicy Binding
  *   System Group
  *   System Group to System Cmd Binding
  *   System Group to System User Binding
  *   System Parameter
  *   System User extended to cover all possible parameters of Netscaler (tested on >= FW v.54.9)

 -- vlazarenko <vlazarenko@squeeze.par>  Wed, 1 Oct 2015 14:06:05 +0100

 nsnitro (1.0.33) unstable; urgency=low

  * New pypi build

 -- vlazarenko <vlazarenko@squeeze.par>  Wed, 31 Mar 2015 09:53:05 +0100

nsnitro (1.0.32) unstable; urgency=low

  * Removed namespace declaration to prevent import warnings.

 -- vlazarenko <vlazarenko@squeeze.par>  Wed, 19 Mar 2014 09:53:05 +0100

nsnitro (1.0.31) unstable; urgency=low

  * NSLBVserverServiceBinding fix

 -- Vladimir Lazarenko <vllazarenko@ebay.com>  Thu, 24 Apr 2014 14:30:52 +0200

nsnitro (1.0.29) unstable; urgency=low

  * Various fixes, including that of logout for 10.x

 -- Vladimir Lazarenko <vllazarenko@ebay.com>  Fri, 18 Apr 2014 18:30:52 +0200

nsnitro (1.0.28) unstable; urgency=low

  * Added delay and graceful options to nsnitrocmd.

 -- mdevreugd <mdevreugd@ebay.com>  Wed, 02 Apr 2014 08:05:12 +0100

nsnitro (1.0.27) unstable; urgency=low

  * Latest updates.

 -- vlazarenko <vlazarenko@squeeze.par>  Thu, 27 Feb 2014 13:28:56 +0100

nsnitro (1.0.26) unstable; urgency=low

  * Latest updates.

 -- vlazarenko <vlazarenko@squeeze.par>  Mon, 16 Dec 2013 10:53:56 +0100

nsnitro (1.0.25) unstable; urgency=low

  * Multiple contributions from various authors

 -- Vladimir Lazarenko <favoretti@gmail.com>  Tue, 12 Nov 2013 01:10:24 +0200

nsnitro (1.0.24) unstable; urgency=low

  * Added ACL resources

 -- Metin de Vreugd <mdevreugd@ebay.com>  Tue, 23 Apr 2013 13:07:24 +0200

nsnitro (1.0.23) unstable; urgency=low

  * SSL certificates linking/unlinking.

 -- Vladimir Lazarenko <vllazarenko@ebay.com>  Fri, 05 Apr 2013 16:03:26 +0200

nsnitro (1.0.22) unstable; urgency=low

  * PEP8 changes here and there, NSSSLCertKey, NSSSLVServerSSLCertKeyBinding

 -- vlazarenko <vllazarenko@ebay.com>  Wed, 20 Feb 2013 23:30:54 +0100

nsnitro (1.0.21) unstable; urgency=low

  * Added NSRewriteAction

 -- Vladimir Lazarenko <vllazarenko@ebay.com>  Tue, 06 Feb 2013 16:24:31 +0100

nsnitro (1.0.20) unstable; urgency=low

  * Added NSFeature management

 -- Vladimir Lazarenko <vllazarenko@ebay.com>  Tue, 05 Feb 2013 16:24:31 +0100

nsnitro (1.0.18) unstable; urgency=low

  * Added VLAN and IP-related classes

 -- vlazarenko <vllazarenko@ebay.com>  Mon, 28 Jan 2013 11:46:53 +0100

nsnitro (1.0.17) unstable; urgency=low

  * --getlbvserverstatusfull command was added to nsnitrocmd.py ('show lb vserver status with bound services')

 -- Vladimir Lazarenko <vllazarenko@ebay.com> Tue, 22 Jan 2013 17:14:41 +0100

nsnitro (1.0.16) unstable; urgency=low

  * Fix nshanode

 -- Vladimir Lazarenko <vllazarenko@ebay.com>  Wed, 03 Oct 2012 03:26:35 +0200

nsnitro (1.0.15) unstable; urgency=low

  * Added NSHANode

 -- Vladimir Lazarenko <vllazarenko@ebay.com>  Tue, 02 Oct 2012 18:01:36 +0200

nsnitro (1.0.14) unstable; urgency=low

  * Added NSHANode
  *

 -- Vladimir Lazarenko <vllazarenko@ebay.com>  Tue, 02 Oct 2012 17:59:36 +0200

nsnitro (1.0.9) unstable; urgency=low

  * Fixed pypi packaging, moved nsnitrocmd.py from contrib to bin and from
    data_files to scripts in setup.py

 -- root <root@debian.localdomain>  Fri, 26 Oct 2012 07:58:59 -0400

nsnitro (1.0.9) unstable; urgency=low

  * Fixed pypi packaging, moved nsnitrocmd.py from contrib to bin and from
    data_files to scripts in setup.py

 -- root <root@debian.localdomain>  Fri, 26 Oct 2012 07:58:59 -0400

nsnitro (1.0.9) unstable; urgency=low

  * Latest contributor's patches
  *    Add responder actions, cs vserver <-> responder policy bindings, add responder policy support etc [timl]
  *    Added default of False to the useSSL option in NSNitro Class. [James Boylan]
  *    LB vserver and service can be added using nsnitrocmd.py [dzhdanov]

 -- root <root@debian.localdomain>  Fri, 26 Oct 2012 07:58:59 -0400

nsnitro (1.0.8) unstable; urgency=low

  * Added delayed service shutdowns

 -- vlazarenko <vlazarenko@squeeze.local>  Sat, 15 Sep 2012 05:08:16 +0200

nsnitro (1.0.7) unstable; urgency=low

  * Working version bump

 -- vlazarenko <vlazarenko@squeeze.local>  Sat, 26 Nov 2011 15:40:17 +0100

nsnitro (1.0.6) unstable; urgency=low

  * Fixed python imports, maybe be backwards-incompatible

 -- vlazarenko <vlazarenko@squeeze.local>  Sat, 26 Nov 2011 15:40:17 +0100

nsnitro (1.0.5) unstable; urgency=low

  * Addded servicegroup support, refactored some code with help of co-contributors

 -- vlazarenko <vlazarenko@squeeze.local>  Sat, 26 Nov 2011 15:40:17 +0100

nsnitro (1.0.4) unstable; urgency=low

  * Adapting for new package structure

 -- vlazarenko <vlazarenko@squeeze.local>  Sat, 26 Nov 2011 15:40:17 +0100

nsnitro (1.0.2) unstable; urgency=low

  * Version bump

 -- vlazarenko <vlazarenko@squeeze.local>  Sat, 26 Nov 2011 15:40:17 +0100

nsnitro (1.0.0) unstable; urgency=low

  * First release.

 -- vlazarenko <vlazarenko@squeeze.local>  Fri, 04 Nov 2011 18:01:24 +0100

nsnitro (0.0.2) unstable; urgency=low

  * Tryout for new structure.
  *

 -- vlazarenko <vlazarenko@squeeze.local>  Fri, 04 Nov 2011 13:46:11 +0100

nsnitro (0.0.1) unstable; urgency=low

  * Initial release.

 -- Vladimir Lazarenko <vladimir@vlfn.nl>  Tue, 08 Nov 2011 23:45:48 +0100


example.py
import time
from nsnitro import *

nitro = NSNitro('localhost', 'api_user', 'api_user')
nitro.login()

# add server test
addserver = NSServer()
addserver.set_name("mp-nitroserver")
addserver.set_ipaddress("10.32.110.99")
NSServer.add(nitro, addserver)

server = NSServer()
server.set_name("mp-nitroserver")
server = server.get(nitro, server)
print server.get_name() + ": " + server.get_state()

# disable server test
server = NSServer()
server.set_name("mp-nitroserver")
NSServer.disable(nitro, server)

time.sleep(2)

server = NSServer()
server.set_name("mp-nitroserver")
server = server.get(nitro, server)
print server.get_name() + ": " + server.get_state()


# enable server test
server = NSServer()
server.set_name("mp-nitroserver")
NSServer.enable(nitro, server)

time.sleep(2)

server = NSServer()
server.set_name("mp-nitroserver")
server = server.get(nitro, server)
print server.get_name() + ": " + server.get_state()

# add service test

addservice = NSService()
addservice.set_name("aurora_testnitroadd")
addservice.set_servername("mp-nitroserver")
addservice.set_servicetype("HTTP")
addservice.set_port(11111)
NSService.add(nitro, addservice)

# add lbvserver test
lbvserver = NSLBVServer()
lbvserver.set_name("nitro_lbvserver_test")
lbvserver.set_ipv46("10.32.110.55")
lbvserver.set_port(11111)
lbvserver.set_clttimeout(180)
lbvserver.set_persistencetype("NONE")
lbvserver.set_servicetype("HTTP")
NSLBVServer.add(nitro, lbvserver)
#
print "LB vserver added"

# bind service to lbvserver test
lbbinding = NSLBVServerServiceBinding()
lbbinding.set_name("nitro_lbvserver_test")
lbbinding.set_servicename("aurora_testnitroadd")
lbbinding.set_weight(40)
NSLBVServerServiceBinding.add(nitro, lbbinding)

print "Binding added"

lbbinding = NSLBVServerServiceBinding()
lbbinding.set_name("nitro_lbvserver_test")
lbbindings = NSLBVServerServiceBinding.get(nitro, lbbinding)

for lbb in lbbindings:
        print "sgn: " + lbb.get_servicegroupname()

# delete binding test
lbbinding = NSLBVServerServiceBinding()
lbbinding.set_name("nitro_lbvserver_test")
lbbinding.set_servicename("aurora_testnitroadd")
NSLBVServerServiceBinding.delete(nitro, lbbinding)

print "Binding removed"

# delete lbvserver test
lbvserver = NSLBVServer()
lbvserver.set_name("nitro_lbvserver_test")
NSLBVServer.delete(nitro, lbvserver)

# get service test

service = NSService()
service.set_name("aurora_testnitroadd")
service = service.get(nitro, service)
print service.get_name() + ": " + service.get_svrstate()
print service.get_name() + ": %s %s" % (service.get_port(), service.get_useproxyport())

#update service test
updateservice = NSService()
updateservice.set_name("aurora_testnitroadd")
updateservice.set_comment("test comment")
updateservice.set_useproxyport("NO")
NSService.update(nitro, updateservice)

# get service test

service = NSService()
service.set_name("aurora_testnitroadd")
service = service.get(nitro, service)
print service.get_name() + ": " + service.get_svrstate()
print service.get_name() + ": %s %s %s" % (service.get_port(), service.get_comment(), service.get_useproxyport())

# disable service test

disservice = NSService()
disservice.set_name("aurora_testnitroadd")
NSService.disable(nitro, disservice)

service = NSService()
service.set_name("aurora_testnitroadd")
service = service.get(nitro, service)
print service.get_name() + ": " + service.get_svrstate()

# enable service test

enservice = NSService()
enservice.set_name("aurora_testnitroadd")
NSService.enable(nitro, enservice)

service = NSService()
service.set_name("aurora_testnitroadd")
service = service.get(nitro, service)
print service.get_name() + ": " + service.get_svrstate()


# rename service test

renservice = NSService()
renservice.set_name("aurora_testnitroadd")
renservice.set_newname("aurora_testnitroadd_rename")
NSService.rename(nitro, renservice)

# rename service back test

renservice = NSService()
renservice.set_name("aurora_testnitroadd_rename")
renservice.set_newname("aurora_testnitroadd")
NSService.rename(nitro, renservice)

# delete service test

delservice = NSService()
delservice.set_name("aurora_testnitroadd")
NSService.delete(nitro, delservice)

try:
        service = NSService()
        service.set_name("aurora_testnitroadd")
        service = service.get(nitro, service)
        print service.get_name() + ": " + service.get_svrstate()
except NSNitroError, e:
        print e.message


# delete server test
delserver = NSServer()
delserver.set_name("mp-nitroserver")
NSServer.delete(nitro, delserver)

try:
        server = NSServer()
        server.set_name("mp-nitroserver")
        server = server.get(nitro, server)
        print server.get_name() + ": " + server.get_state()
except NSNitroError, e:
        print e.message


# add a snmp community string
addsnmpstring = SNMPCommunity()
addsnmpstring.set_communityname('nitroExample')
addsnmpstring.set_permissions('GET')

SNMPCommunity.add(nitro, addsnmpstring)

# get the community string
snmpstring = SNMPCommunity.get(nitro, 'nitroExample')

# get all the community strings
snmpstrings = SNMPCommunity.get_all(nitro)

# delete the community string that was initially added
SNMPCommunity.delete(nitro, addsnmpstring)

# Add a test user
test_user = SystemUser({'username' : 'nsnitro-example', 'password' : 'test123'})
SystemUser.add(nitro, test_user)

# Get all the users
users = SystemUser.get_all(nitro)

# Get a user
user = SystemUser.get(nitro, test_user)

# Update the password of a user
SystemUser.update(nitro, SystemUser({'username' : 'nsnitro-example', 'password' :'321test'}))

# Delete the test user
SystemUser.delete(nitro, test_user)


# Get the ns version
nsversion = NSVersion().get(nitro)
